integrations/tensorflow/test/tfe2e/math/math_test.py

```diff
@@ -0,0 +1,751 @@
+
+
+def _wrap_top_k(top_k):
+  # top_k returns a tensorflow.python.ops.gen_nn_ops.TopKV2. Wrap it in a tuple
+  # so we don't have to. (The lambda is wrapped to avoid a recursive capture).
+  return lambda *args, **kwargs: tuple(top_k(*args, **kwargs))
+
+
+def create_function_unit_test(
+    function_name: str,
+    unit_test_spec: tf_test_utils.UnitTestSpec) -> tf.function:
+  """Creates a tf_function_unit_test from the provided UnitTestSpec."""
+  function = getattr(tf.math, function_name)
+  signature = unit_test_spec.input_signature
+
+  if tf_utils.is_complex(signature):
+    function, signature = tf_utils.rewrite_complex_signature(
+        function, signature)
+  if function_name == "top_k":
+    function = _wrap_top_k(function)
+  wrapped_function = lambda *args: function(*args, **unit_test_spec.kwargs)
+
+  if FLAGS.dynamic_dims:
+    signature = tf_utils.apply_function(signature, tf_utils.make_dims_dynamic)
+
+  return tf_test_utils.tf_function_unit_test(
+      input_signature=signature,
+      input_generator=unit_test_spec.input_generator,
+      input_args=unit_test_spec.input_args,
+      name=unit_test_spec.unit_test_name,
+      rtol=1e-5,
+      atol=1e-5)(wrapped_function)
+
+
+class TfMathModule(tf_test_utils.TestModule):
+
+  def __init__(self):
+    super().__init__()
+    for function in FLAGS.functions:
+      for unit_test_spec in FUNCTIONS_TO_UNIT_TEST_SPECS[function]:
+        if not FLAGS.test_complex and tf_utils.is_complex(
+            unit_test_spec.input_signature):
+          continue
+        function_unit_test = create_function_unit_test(function, unit_test_spec)
+        setattr(self, unit_test_spec.unit_test_name, function_unit_test)
+
+
+def get_relative_artifacts_dir() -> str:
+  if len(FLAGS.functions) > 1:
+    # We only allow testing multiple functions with a single target backend
+    # so that we can store the artifacts under:
+    #   'artifacts_dir/multiple_functions__backend/...'
+    # We specialize the 'multiple_functions' dir by backend to avoid overwriting
+    # tf_input.mlir and iree_input.mlir. These are typically identical across
+    # backends, but are not when the functions to compile change per-backend.
+    if len(FLAGS.target_backends) != 1:
+      raise flags.IllegalFlagValueError(
+          "Expected len(target_backends) == 1 when len(functions) > 1, but got "
+          f"the following values for target_backends: {FLAGS.target_backends}.")
+    function_str = f"multiple_functions__{FLAGS.target_backends[0]}"
+  else:
+    function_str = FLAGS.functions[0]
+  dim_str = "dynamic_dims" if FLAGS.dynamic_dims else "static_dims"
+  complex_str = "complex" if FLAGS.test_complex else "non_complex"
+  return os.path.join("tf", "math", function_str, f"{dim_str}_{complex_str}")
+
+
+class TfMathTest(tf_test_utils.TracedModuleTestCase):
+
+  def __init__(self, *args, **kwargs):
+    super().__init__(*args, **kwargs)
+    self._modules = tf_test_utils.compile_tf_module(
```

Comment:
Could we / do we want to eventually make these sorts of tests pause after importing from the frontend?

1. import from python to .mlir (requires a build of the import tool on the host machine)
2. compile from .mlir to .vmfb (requires a build of IREE's generic compiler on the host machine)
3. run the compiled tests and compare against the reference backend(s)

I'd like to be able to do (2) and (3) on my Windows dev machine, and I imagine we'd also like to do (1) and (2) on Buildkite for running (3) on Android.

---

This generator approach seems to get us closer to that, but I don't quite see the full picture yet.

Reply:
Yeah, definitely a lot more to be done at the runner level once we get this in place. I basically set it up to emulate/use the existing lower level tests so we can get to coverage, but a follow-on step needs to rip that up and come up with a better runner/test that is disaggregated as you say. My one bow towards that is that I named the one current runner "tfhost" and expect to add another/refactor the scripts to get where you are talking about.